homeassistant/components/nzbget/__init__.py

```diff
@@ -52,147 +55,213 @@
-        nzbget_api.version()
-    except pynzbgetapi.NZBGetAPIException as conn_err:
-        _LOGGER.error("Error setting up NZBGet API: %s", conn_err)
-        return False
+    if DOMAIN in config:
+        hass.async_create_task(
+            hass.config_entries.flow.async_init(
+                DOMAIN, context={"source": SOURCE_IMPORT}, data=config[DOMAIN],
+            )
+        )
 
-    _LOGGER.debug("Successfully validated NZBGet API connection")
+    return True
 
-    nzbget_data = hass.data[DATA_NZBGET] = NZBGetData(hass, nzbget_api)
-    nzbget_data.init_download_list()
-    nzbget_data.update()
 
-    def service_handler(service):
-        """Handle service calls."""
-        if service.service == SERVICE_PAUSE:
-            nzbget_data.pause_download()
-        elif service.service == SERVICE_RESUME:
-            nzbget_data.resume_download()
-        elif service.service == SERVICE_SET_SPEED:
-            limit = service.data[ATTR_SPEED]
-            nzbget_data.rate(limit)
+async def async_setup_entry(hass: HomeAssistantType, entry: ConfigEntry) -> bool:
+    """Set up NZBGet from a config entry."""
+    if not entry.options:
+        options = {
+            CONF_SCAN_INTERVAL: entry.data.get(
+                CONF_SCAN_INTERVAL, DEFAULT_SCAN_INTERVAL
+            ),
+        }
+        hass.config_entries.async_update_entry(entry, options=options)
 
-    hass.services.register(
-        DOMAIN, SERVICE_PAUSE, service_handler, schema=vol.Schema({})
+    coordinator = NZBGetDataUpdateCoordinator(
+        hass, config=entry.data, options=entry.options,
     )
 
-    hass.services.register(
-        DOMAIN, SERVICE_RESUME, service_handler, schema=vol.Schema({})
+    await coordinator.async_refresh()
+
+    if not coordinator.last_update_success:
+        raise ConfigEntryNotReady
+
+    undo_listener = entry.add_update_listener(_async_update_listener)
+
+    hass.data[DOMAIN][entry.entry_id] = {
+        DATA_COORDINATOR: coordinator,
+        DATA_UNDO_UPDATE_LISTENER: undo_listener,
+    }
+
+    for component in PLATFORMS:
+        hass.async_create_task(
+            hass.config_entries.async_forward_entry_setup(entry, component)
+        )
+
+    def pause(call) -> None:
+        """Service call to pause downloads in NZBGet."""
+        coordinator.nzbget.pausedownload()
+
+    def resume(call) -> None:
+        """Service call to resume downloads in NZBGet."""
+        coordinator.nzbget.resumedownload()
+
+    def set_speed(call) -> None:
+        """Service call to rate limit speeds in NZBGet."""
+        coordinator.nzbget.rate(call.data[ATTR_SPEED])
+
+    hass.services.async_register(DOMAIN, SERVICE_PAUSE, pause, schema=vol.Schema({}))
+
+    hass.services.async_register(DOMAIN, SERVICE_RESUME, resume, schema=vol.Schema({}))
+
+    hass.services.async_register(
+        DOMAIN, SERVICE_SET_SPEED, set_speed, schema=SPEED_LIMIT_SCHEMA
     )
```

Comment:
Personal preference for me would be to break this up into `_async_register_services(coordinator)`